### PR TITLE
HDDS-9013. Fetch root CA certificate list during SCM startup.

### PR DESCRIPTION
**Pull request by ChenSammi**
https://issues.apache.org/jira/browse/HDDS-9013

### CODE REVIEW

**Comment by fapifta** [comment]
Hi @ChenSammi, 

thank you for working on this, I have a fundamental problem with publishing the notifyNotificationReceivers method. With that any class can initiate sending a notification about refreshed certificates, but in my mind, the internal handling of the rootCA certificate, and the actual service's certificate is something we would like to have handled inside the certificate client.

I have added a PR to your branch, to showcase how I would solve the problem, with extending the SCMCertificateClient instead of the SCMHAManager with the task of refreshing the rootCA certificate. Please take a look, and let me know what do you think.

**Comment by ChenSammi** [comment]
@fapifta , I didn't find the patch you mentioned on [ChenSammi:HDDS-9013](https://github.com/ChenSammi/ozone/tree/HDDS-9013) branch. Could you check it?

**Comment by Galsza** [comment]
Hi Sammi, here is the patch Pifta mentioned: https://github.com/apache/ozone/commit/8bdc184742defa5516af1560e825de2d13eddb36


Overall your changes are pretty similar to it now. My question is: do we need to run refreshCaCertificates through the executorService? (Do we need to run it in sync?)

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMHAManagerImpl.java [hunk 1]
```diff
@@ -18,6 +18,7 @@
 package org.apache.hadoop.hdds.scm.ha;
 
 import com.google.common.base.Preconditions;
+
```
Comment (fapifta) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMHAManagerImpl.java:21]: Nit:
if we touch imports here, probably we should move TimeUnit import to java.* imports, otherwise I would just not add a newline here. (Please ignore this if no other changes are done on the PR).

File: hadoop-ozone/dist/src/main/compose/testlib.sh [hunk 2]
```diff
@@ -278,7 +278,7 @@ save_container_logs() {
   local output_name=$(get_output_name)
   local c
   for c in $(docker-compose ps "$@" | cut -f1 -d' ' | tail -n +3); do
-    docker logs "${c}" &>> "$RESULT_DIR/docker-${output_name}${c}.log"
+    docker logs "${c}" >> "$RESULT_DIR/docker-${output_name}${c}.log" 2>&1
```
Comment (fapifta) [inline hadoop-ozone/dist/src/main/compose/testlib.sh:281]: I am not sure what is actually fixed here, but probably this is unrelated to this patch, and should be fixed separately.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/SCMCertificateClient.java [hunk 3]
```diff
@@ -228,4 +235,69 @@ public String signAndStoreCertificate(PKCS10CertificationRequest request,
...
+        }
+
+        for (X509Certificate cert : rootCAsFromLeaderSCM) {
+          LOG.info("Fetched new root CA certificate {} from leader SCM",
+              cert.getSerialNumber().toString());
+          storeCertificate(
+              CertificateCodec.getPEMEncodedString(cert), CAType.SUBORDINATE);
+        }
+        String scmCertId = getCertificate().getSerialNumber().toString();
+        notifyNotificationReceivers(scmCertId, scmCertId);
```
Comment (fapifta) [inline hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/SCMCertificateClient.java:288]: We are using the same certificate serial id as old and new in the notification. Is this intentional, if so can you please elaborate I do not see why we do so.

**Review by fapifta — COMMENTED** [review]
Thank you @ChenSammi  for taking a look on my suggested approach also for adding further improvements. I have a few comments inline, no big things, but I thought it is better to discuss at least before we commit, please take a look.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMHAManagerImpl.java (hunk 1 above)
Comment (ChenSammi) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SCMHAManagerImpl.java:21]: I will revert it.

File: hadoop-ozone/dist/src/main/compose/testlib.sh (hunk 2 above)
Comment (ChenSammi) [reply hadoop-ozone/dist/src/main/compose/testlib.sh:281]: Make sense.

**Comment by ChenSammi** [comment]
> Overall your changes are pretty similar to it now. My question is: do we need to run refreshCaCertificates through the executorService? (Do we need to run it in sync?)

refreshCaCertificates is required to run parallelly with the SCM main Ratis server and ratis client involved actions, so that it will solve the case that SCM cannot join the ratis group, or ratis client cannot connect to retis server, because of lacking new root certificates.

File: hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/SCMCertificateClient.java (hunk 3 above)
Comment (ChenSammi) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/SCMCertificateClient.java:288]: It's elaborate. Because the interface API CertificateNotification#notifyCertificateRenewed requires two certificate ID as the input parameters.

Comment (fapifta) [reply hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/SCMCertificateClient.java:288]: As this is pretty much internal within SCM, I am fine with this approach. If it would be a more open and potentially more frequently used API I would argue that in this case we should set the old certificate id to null for the notification, but it is absolutely sufficient for now, and looking at it a bit deeper, within SCM we don't even use these values, the are important in the logic of the delegation token generation.

**Comment by Galsza** [comment]
@ChenSammi Thank you for explaining. The changes are looking good to me now.
